### ISSUE TITLE
Off by one errors in Perl_utf16_to_utf8()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9914]
Migrated from [rt.perl.org#69859](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=69859) (status was 'resolved')

Searchable as RT69859$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



Consider the character U\-10FC00\, whose surrogate pairs are DBFF DC00

$ \./perl \-Ilib \-MEncode \-we 'print "\\xFE\\xFF"\, encode\("UTF\-16BE"\, "qq\[\\x\{10fc00\}\]"\)' >script\.pl

$ \./perl script\.pl
Malformed UTF\-8 character \(UTF\-16 surrogate 0xdbff\) at script\.pl line 1\.
Malformed UTF\-8 character \(UTF\-16 surrogate 0xdc00\) at script\.pl line 1\.

This is wrong\. As is

$ \./perl \-Ilib \-MEncode \-we 'print "\\xFE\\xFF"\, encode\("UTF\-16BE"\, "qq\[\\x\{103ff\}\]"\)' >script\.pl

$ \./perl script\.pl Malformed UTF\-16 surrogate\.


In both cases this is silent&#8203;:

$ iconv \-f UTF\-16BE \-t UTF\-8 \<script\.pl | \./perl

There are two off\-by\-one error in the surrogate handling in
Perl\_utf16\_to\_utf8\(\)\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.11.0.42:

Configured by nick at Fri Oct 16 20:36:59 BST 2009.

Summary of my perl5 (revision 5 version 11 subversion 0) configuration:
  Commit id: 20d0b1e9c410d995ea730a00781152c652d4b672
  Platform:
    osname=linux, osvers=2.6.18-xenu, archname=x86_64-linux-thread-multi
    uname='linux zazen 2.6.18-xenu #1 smp thu oct 4 12:23:41 bst 2007 x86_64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache g++ -Dld=g++ -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -Doptimize=-Os -Dusethreads -Duse64bitall -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache g++', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-Os',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='g++', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc.so.6, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -Os -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.11.0.42:
    lib
    /home/nick/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e/lib/perl5/site_perl/5.11.0/x86_64-linux-thread-multi
    /home/nick/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e/lib/perl5/site_perl/5.11.0
    /home/nick/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e/lib/perl5/5.11.0/x86_64-linux-thread-multi
    /home/nick/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e/lib/perl5/5.11.0
    .


Environment for perl 5.11.0.42:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Sat Oct 17 12&#8203;:51&#8203;:28 2009\, nicholas wrote&#8203;:

> $ \./perl \-Ilib \-MEncode \-we 'print "\\xFE\\xFF"\, encode\("UTF\-16BE"\,
>    "qq\[\\x\{10fc00\}\]"\)' >script\.pl

No\, RT\, "nicholas" did not write \*that\*\. You've inserted some line
breaks for me\, that mess up the sense of it\.

> There are two off\-by\-one error in the surrogate handling in
> Perl\_utf16\_to\_utf8\(\)\.

I fixed these in 52b9aa85a8c28ddc591b0a7f2b1f8c729075d9a1

Nicholas Clark

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'new' to 'resolved'
